src/t_hash.c

```diff
@@ -2267,29 +2965,976 @@ void hpersistCommand(client *c) {
+    addReplyError(c, "Only one of NX, XX, GT, and LT arguments can be specified");
+    return C_ERR;
+err_expiration:
+    addReplyError(c, "Only one of EX, PX, EXAT, PXAT or KEEPTTL arguments can be specified");
+    return C_ERR;
+err_field_condition:
+    addReplyError(c, "Only one of DCF or DOF arguments can be specified");
+    return C_ERR;
+err_return_condition:
+    addReplyError(c, "Only one of GETOLD or GETNEW arguments can be specified");
+    return C_ERR;
+}
+
+/*
+ * Set field value and optionally set the field's remaining time to live.
+ * Optionally it creates the key/fields.
+ *
+ * HSETF key
+ *      [DC] [DCF | DOF]
+ *      [NX | XX | GT | LT]
+ *      [GETNEW | GETOLD]
+ *      [EX seconds | PX milliseconds | EXAT unix-time-seconds | PXAT unix-time-milliseconds | KEEPTTL]
+ *      <FVS count field value [field value …]>
+ */
+void hsetfCommand(client *c) {
+    int flags = 0;
+    robj *hashObj, *keyArg = c->argv[1];
+
+    int firstFieldPos = 0;
+    int numFields = 0;
+    uint64_t expireAt = EB_EXPIRE_TIME_INVALID;
+
+    if (hsetfParseArgs(c, &flags, &expireAt, &firstFieldPos, &numFields) != C_OK)
+        return;
+
+    hashObj = lookupKeyWrite(c->db, c->argv[1]);
+    if (!hashObj) {
+        /* Don't create object if command has DC argument */
+        if (flags & HFE_CMD_DC) {
+            addReplyOrErrorObject(c, shared.null[c->resp]);
+            return;
         }
+        hashObj = createHashObject();
+        dbAdd(c->db,c->argv[1],hashObj);
+    }
+
+    hashTypeTryConversion(c->db,hashObj,c->argv,
+                          firstFieldPos,
+                          firstFieldPos + (numFields * 2) - 1);
+
+    attachHfeMeta(c->db, hashObj, keyArg);
+    uint64_t minExpire = hashTypeGetMinExpire(hashObj);
+
+    /* Figure out from provided set of fields in command, which one has the minimum
+     * expiration time, before the modification (Will be used for optimization below) */
+    uint64_t minExpireFields = EB_EXPIRE_TIME_INVALID;
+
+    int updated = 0;
+    addReplyArrayLen(c, numFields);
+    for (int i = 0; i < numFields ; i++) {
+        sds field = c->argv[firstFieldPos + (i * 2)]->ptr;
+        sds value = c->argv[firstFieldPos + (i * 2) + 1]->ptr;
+        updated += hsetfSetFieldAndReply(c, hashObj, field, value, flags,
+                                         expireAt, &minExpireFields);
+    }
 
-        hfieldPersist(c->db, hashObj, hf);
-        addReplyLongLong(c, HFE_PERSIST_OK);
+    /* Notify keyspace event, update dirty count and update global HFE DS */
+    if (updated > 0) {
+        server.dirty += updated;
+        signalModifiedKey(c,c->db,keyArg);
+        notifyKeyspaceEvent(NOTIFY_HASH,"hsetf",keyArg,c->db->id);
+        if (hashTypeLength(hashObj, 0) == 0) {
+            dbDelete(c->db,keyArg);
+            notifyKeyspaceEvent(NOTIFY_GENERIC,"del",keyArg, c->db->id);
+        } else {
+            updateGlobalHfeDs(c->db, hashObj, minExpire, minExpireFields);
+        }
     }
```

Comment:
if updated is 0, we should check if hashObj is empty and delete it to avoid create a empty hash.

Reply:
So the question is how do we handle the case:
- We create the object as DC is not provided.
- hsetf does not set any field
- We delete the empty object

In this case, we'll be sending notifications for empty object.

Reply:
ping @oranagra, in this case, we will send two notifications(new, del), do you have any advices?

Reply:
```
if (!hashObj) {
    ......
        /* If object does not exist and DCF flag is given, no need to create
         * the object as we won't create any field. */
        if (flags & HFE_CMD_DCF) {
            addReplyArrayLen(c, numFields);
            for (int i = 0; i < numFields ; i++) {
                if (flags & (HFE_CMD_GETOLD | HFE_CMD_GETNEW))
                    addReplyNull(c);
                else
                    addReplyLongLong(c, HSETF_FAIL);
            }
            return;
        }
```

Added this early check to avoid creation of the object. A bit ugly but no way around it.

Reply:
@tezc i would prefer to send 2 notifications, which is pretty rare, unless someone intentionally sends error command.

Reply:
I don't have strong opinion but pinged Oran and he also asked if we can do an early check to avoid it. Well, maybe this early check is not worth it. Not sure.

Reply:
not sure the following sulution is avaliable, since we might need to update the fields ebuckets and db->hexpires.
btw: should we update db->hexpires ebuckets after dbDelete()?
```
if (!hashObj) {
    created = 1;
    createHashObject();
}

...

if (updated == 0) {
    if (created && hashTypeLength(hashObj, 0) == 0)
            decrRefCount(hashObj);
} else {
    dbAdd();
}
```

Reply:
To do this, we also need to do `attachHfeMeta()` lazily. Inside of it, we use the key reference in dict. It becomes more complicated imho. Also, it will be different than other commands, it feel like it might be confusing. 

> btw: should we update db->hexpires ebuckets after dbDelete()?

Hope I understand this question correctly, if `hashTypeLength(hashObj, 0) == 0` then, object is not part of db->hexpires. So, dbDelete() should be safe.

Reply:
Alternatively, we can add DCF check next to DC and return null rather than replying to all fields. 

```
        /* Don't create object if command has DC argument */
        if (flags & HFE_CMD_DC || flags & HFE_CMD_DCF) {
            addReplyOrErrorObject(c, shared.null[c->resp]);
            return;
        }
```

Reply:
yeah, i like it.